### PR TITLE
[SHARE-510][Feature] Archive CeleryTasks

### PR DESCRIPTION
**Pull request by leb2dg**
## Purpose
Remove old celery tasks from database and store on Amazon s3.

## Changes
Added ArchiveBot that compresses a ~~CSV~~ *JSON* version of old celery tasks, sends those files to s3, then deletes the tasks from the database.

## Side effects
Tasks will be removed after set period of time.

### CODE REVIEW

File: bots/archive/bot.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        self.archive_tasks = CeleryProviderTask.objects.filter(
+            name='share.tasks.ArchiveTask',
+            timestamp__lt=current_time + two_weeks
+        )
+
+        if self.elasticsearch_tasks.exists():
+            compressed_data = self.queryset_to_compressed_csv(self.elasticsearch_tasks)
+            self.put_s3(bucket, 'elasticsearch/elasticsearch_tasks_', compressed_data)
+
+            logger.info('Finished archiving data for Elasticsearch CeleryTask at %s', datetime.datetime.utcnow().isoformat())
```
Comment (leb2dg) [inline bots/archive/bot.py]: Ohh, awk. I was just formatting the logs like the ones in `DisambiguatorTask`, `ApplySingleChangeSet`, and `ApplyChangeSets` in [tasks.py](https://github.com/CenterForOpenScience/SHARE/blob/develop/share/tasks.py#L194).

Comment (chrisseto) [inline bots/archive/bot.py]: Huh, in that case could you double check if logger statements do or do not include timestamps?

Comment (leb2dg) [inline bots/archive/bot.py]: Logs do have timestamps 👍

File: bots/archive/bot.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            current_date = datetime.datetime.utcnow().isoformat()
+            s3.Object(bucket, location + current_date + '.gz').put(Body=data.getvalue())
+        except botocore.exceptions.ClientError as e:
+            raise botocore.exceptions.ClientError(e)
+
+    def delete_queryset(self, querySet):
+        num_deleted = 0
+        try:
+            with transaction.atomic():
+                num_deleted = querySet.delete()
```
Comment (leb2dg) [inline bots/archive/bot.py]: You recall correctly

File: bots/archive/bot.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            l = []
+            for _ in range(size):
+                l.append(next(iterable))
+            yield l
+    except StopIteration:
+        yield l
+
+
+def get_bucket(bucket_name):
+        s3 = boto3.resource('s3')
```
Comment (chrisseto) [inline bots/archive/bot.py]: Double indented?

File: bots/archive/bot.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            s3.Object(bucket, location + current_date + '.bz2').put(Body=data.getvalue())
+        except botocore.exceptions.ClientError as e:
+            raise botocore.exceptions.ClientError(e)
+
+
+def delete_queryset(queryset):
+        num_deleted = 0
+        try:
+            with transaction.atomic():
+                (num_deleted, deleted_metadata) = queryset.delete()
```
Comment (chrisseto) [inline bots/archive/bot.py]: Don't need the `()` here.

File: bots/archive/bot.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                'ArchiveTask'
+            )
+
+    def archive_queryset(self, queryset, location, task_name):
+        total = queryset.count()
+        logger.info('Found %s %ss eligible for archiving', total, task_name)
+        logger.info('Archiving in chunks of %d', self.chunk_size)
+
+        model = queryset.model
+        for i, celery_tasks in enumerate(chunk(queryset.all(), self.chunk_size)):
```
Comment (chrisseto) [inline bots/archive/bot.py]: Use `queryset.iterator` here, it'll save a lot of memory. `.all()` Loads everything into memory at once.

**Comment by leb2dg** [comment]
Takes about a minute for 10,000 records. Memory usage from ~100 - ~150 MiB

**Comment by chrisseto** [comment]
Last thing (Otherwise looks great) Your serialization is a touch off. Looks like you're just running `__str__` on everything. Somethings need to be decoded (the bytes) or need to be turned from the model to their primary key.

(Ignore the edu.utktrace, that's just my local database not matching up)

Currently:
```
19746350,b'share.celeryprovidertask',620d191d-d66f-4100-a992-758ae53a461b,b'share.tasks.BotTask',b'()',b'{}',2017-01-09 04:19:00.008353+00:00,"<ShareUser(1, system)>","<ShareUser(85, providers.edu.utktrace)>",3,b'elasticsearch',b'0.0.1'
```

Ideally
```
19746350,share.celeryprovidertask,620d191d-d66f-4100-a992-758ae53a461b,share.tasks.BotTask,(),{},2017-01-09 04:19:00.008353+00:00,1,85,3,elasticsearch,0.0.1
```

An easy way to test is by trying to import the file back into your database after exporting it.

**Comment by leb2dg** [comment]
JSON can be loaded back into the db by running `./manage.py loaddata path/to/file.json`

**Comment by chrisseto** [comment]
Looks like there's one tiny issue

```
--- Logging error ---
Traceback (most recent call last):
  File "/Users/chrisseto/.pyenv/versions/3.5.2/lib/python3.5/logging/__init__.py", line 980, in emit
    msg = self.format(record)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/lib/python3.5/logging/__init__.py", line 830, in format
    return fmt.format(record)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/colorlog/colorlog.py", line 129, in format
    message = super(ColoredFormatter, self).format(record)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/lib/python3.5/logging/__init__.py", line 567, in format
    record.message = record.getMessage()
  File "/Users/chrisseto/.pyenv/versions/3.5.2/lib/python3.5/logging/__init__.py", line 330, in getMessage
    msg = msg % self.args
TypeError: %d format: a number is required, not Page
Call stack:
  File "./manage.py", line 17, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/Users/chrisseto/Documents/SHARE/share/share/management/commands/runbot.py", line 43, in handle
    BotTask().apply(task_args, task_kwargs, throw=True)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/celery/app/task.py", line 755, in apply
    request=request, propagate=throw)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/celery/app/trace.py", line 355, in eager_trace_task
    uuid, args, kwargs, request)
  File "/Users/chrisseto/.pyenv/versions/3.5.2/envs/share/lib/python3.5/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/Users/chrisseto/Documents/SHARE/share/share/tasks.py", line 65, in run
    self.do_run(*self.args, **self.kwargs)
  File "/Users/chrisseto/Documents/SHARE/share/share/tasks.py", line 226, in do_run
    bot.run()
  File "/Users/chrisseto/Documents/SHARE/share/bots/archive/bot.py", line 129, in run
    'NormalizerTask'
  File "/Users/chrisseto/Documents/SHARE/share/bots/archive/bot.py", line 172, in paginate_queryset
    logger.info('Archived %d', page)
Message: 'Archived %d'
Arguments: (<Page 2 of 2>,)
```

**Comment by leb2dg** [comment]
How did I never get that?? Will fix 👍

File: bots/archive/bot.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    compressed_output.write(bz2.compress(str.encode(output.getvalue())))
+
+    return compressed_output
+
+
+def put_s3(bucket, location, data):
+    s3 = boto3.resource('s3')
+    try:
+        current_date = datetime.datetime.utcnow().isoformat()
+        s3.Object(bucket, location + current_date + '.bz2').put(Body=data.getvalue())
```
Comment (chrisseto) [inline bots/archive/bot.py]: Final request, I promise. change `.bz2` to `.json.bz2`. Django whines when there is no file extension and the archive utility on OSX defaults file names to the archive name minus `.zip/bz/tar.gz`

**Comment by leb2dg** [comment]
screen shot 2017-02-23 at 11 22 14 am

File: bots/archive/bot.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        current_time = datetime.datetime.utcnow()
+        one_day = datetime.timedelta(days=-1)
+        one_week = datetime.timedelta(weeks=-1)
+        two_weeks = datetime.timedelta(weeks=-2)
+
+        # bots.elasticsearch (24hrs)
+        elasticsearch_tasks = CeleryProviderTask.objects.filter(
+            app_label='elasticsearch',
+            timestamp__lt=current_time + one_day,
+            status=3
```
Comment (chrisseto) [inline bots/archive/bot.py]: `CeleryTask.STATUS.succeeded` or `CeleryProviderTask.STATUS.succeeded`

**Pull request merged**